src/check_nesting.cpp

```diff
@@ -0,0 +1,53 @@
+#include "sass.hpp"
+#include <vector>
+
+#include "check_nesting.hpp"
+#include "context.hpp"
+
+namespace Sass {
+
+  CheckNesting::CheckNesting(Context& ctx)
+  : ctx(ctx),
+    block_stack(std::vector<Block*>())
+  { }
+
+  Statement* CheckNesting::operator()(Block* b)
+  {
+    Block* bb = SASS_MEMORY_NEW(ctx.mem, Block, b->pstate(), b->length(), b->is_root());
+    block_stack.push_back(bb);
+    append_block(b);
+    block_stack.pop_back();
+    return bb;
+  }
+
+  Statement* CheckNesting::operator()(Declaration* d)
+  {
+    if (block_stack.size() < 2) {
```

Comment:
This check [still needs work](https://github.com/sass/sass/blob/63586f1e51c0aec47a7afcc8cd17aa03f32f0a29/lib/sass/tree/visitors/check_nesting.rb#L140-L141) and the test case will need to be updated.